### ISSUE TITLE
lock contention (?) while downloading from the gateway/linksharing with high throughput

### ISSUE THREAD

**Issue by amwolff** [storj/edge#358]
We found out that a single AMD EPYC 7702P can only push out as much as 7.5-10 Gbps. We think that it might be possible to push out at least as much as twice more, but there's some kind of contention happening deep within the runtime (maybe caused by how we use the language?).

Flamegraph: https://pprof.host/qw0g/flamegraph

![pprof host_qw0g_flamegraph](https://github.com/storj/gateway-mt/assets/30234756/e87e5262-eebb-44b5-a695-1f21ed6cc058)

**Comment by azdagron** [comment]
Do we know what version of the Linux kernel was running on that box?

**Comment by azdagron** [comment]
And the version of Go used to compile the gateway-mt binary?

**Comment by halkyon** [comment]
@azdagron 

* Linux 6.2 (we also saw it with 5.19 before it got upgraded on Ubuntu 22.04 HWE)
* Go 1.20.6

**Comment by amwolff** [comment]
@pwilloughby bumped the priority of this. Feel free to grab it.

**Comment by jseaton** [comment]
Okay, I can reproduce this more or less with gateway-st:

https://pprof.host/nc40/flamegraph

Using a 360 thread VM, Ubuntu 20.04.
Generating load with CUNO on the same machine (just a single file copy!).
Will try to remember to measure throughput on next run...

Interestingly I can't so far repro these flame graphs just with `storj-up` *at all* even though I can do a solid 12Gbit on a test server (hits unrelated limits I think).

I can think of two broad reasons this could be true:
- we're not hiding network latency very well
- extra CPU load in production

I'm going to try fiddling with `storj-up` to see if I can make the issue appear there.

**Comment by jseaton** [comment]
Okay, after a bunch of analysis, it looks like:
- yes, I hit a solid performance wall at ~5Gbit
- for comparison I can do about 9Gbit with CUNO to an AWS bucket from the same machine
- yes, there is a real issue with epoll here, I can see the kernel spending a lot of time inside osq_lock in epoll
    - perf graph: ![perf out 2 fold](https://github.com/user-attachments/assets/4d94e129-1524-460f-98ff-6debbb2ff65d) (sorry for the rubbish graph)
- I can see a similar performance wall when running parallel downloads manually from a simple Go program
- this issue is fixed in kernels around 5.11 - epollwait (mostly) goes away
    - pprof: https://pprof.host/pr40/flamegraph
- I also tried a patched version of Go which seems to have a similar effect: https://pprof.host/pw40/flamegraph
- however, I still see around the same performance! It's not clear to me if this is the Go runtime's fault here, but I can still see this with my silly parallel download Go program
- I tried some obvious things like turning off the GC and altering `GOMAXPROCS` (no effect performance-wise)
    - pprof graph without GC, note the big `runtime.findRunnable`: https://pprof.host/pm40/flamegraph
- This all suggests to me that we're hitting a slightly different Go runtime limitation, and just optimising for CPU cycles probably won't help.

The part I still really don't understand is why Storj-up doesn't have this issue. I'm a bit concerned that what I'm really seeing is a limitation in e.g. the satellite, and the Go scheduler appearing in the flamegraphs is just an artefact of Go not having enough work to do.

TL;DR I think ensuring our gateways are running on kernels >= 5.11 and maybe recommending that to users would be wise, but sadly that won't fix the performance issue alone. Patching Go itself may also be worth considering for old kernels, with the same caveat.

**Comment by onionjake** [comment]
Does storj-up use localhost? localhost addresses have a different codepath in the kernel I think.

**Comment by halkyon** [comment]
All gateways are hosted on servers with Linux 5.15 currently, and everything is built with Go 1.23.3. https://github.com/golang/go/issues/65064 might also be interesting.

**Comment by amwolff** [comment]
A few things emerged after my last comment here, and I did a write-up on the latest testing in another issue [here](https://github.com/golang/go/issues/65064#issuecomment-2607813860).
